blockchain/src/history/history_store.rs

```diff
@@ -703,56 +691,30 @@ impl HistoryInterface for HistoryStore {
             }
         };
 
-        // Get history tree for given epoch.
-        let tree = MerkleMountainRange::new(MMRStore::with_read_transaction(
-            &self.hist_tree_table,
-            txn,
-            epoch_number,
-        ));
-
-        // Return early if there are no leaves in the HistoryTree for the given epoch.
-        let num_leaves = tree.num_leaves();
-        if num_leaves == 0 {
-            return 0;
-        }
-
-        // Find the number of the last macro stored for the given epoch.
-        let last_tx = self
-            .get_historic_tx(epoch_number, (num_leaves - 1) as u32, Some(txn))
-            .unwrap();
-        let last_macro_block = Policy::last_macro_block(last_tx.block_number);
-
-        let end = if Policy::epoch_at(last_macro_block) != epoch_number {
-            0
-        } else {
-            // Otherwise, find the last block that had transactions.
-            // Usually that is the macro block itself.
-            // Only in the first batch, we actually need the loop.
-            let mut block_number = last_macro_block;
-            let mut end = 0;
-            loop {
-                let (block_start, block_end) = self.get_indexes_for_block(block_number, Some(txn));
-                // If start != end, this is the last block that contained finalized transactions.
-                if block_start != block_end {
-                    end = block_end;
-                    break;
-                }
-                // If we switched epochs, start at 0.
-                if Policy::epoch_at(block_number) != epoch_number {
-                    break;
-                }
-                block_number -= 1;
+        // Find the index of the last transaction that occurred before the given block.
+        let epoch_number = Policy::epoch_at(block_number);
+        loop {
+            // If start != end, this is the last block that contained transactions.
+            let (start, end) = self.get_indexes_for_block(block_number, Some(txn));
+            if start != end {
+                break end as usize;
             }
-            end
-        };
-
-        end as usize
+            // If we switched epochs, start at 0.
+            if Policy::epoch_at(block_number) != epoch_number {
+                break 0;
+            }
+            // If we are at the genesis block, start at 0.
```

Comment:
_start at 0?_ Maybe:
```suggestion
            // If we are at the genesis block, this epoch is empty.
```
Although this looks like a hack because there are going to be transactions for blocks below the genesis